### PR TITLE
Add lock expiry handler to expire state locks

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add lock expiry handler to expire state locks

## Motivation and Context
Add lock expiry handler to expire state locks

## How to test this PR?
This is a historical code bringing in back such that 
we don't loose on the historical functionality.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#53c84b40-b36e-4f7e-92d5-8baea812ae20) |
| mint-large-bucket.sh | :x: [more...](#1f83be75-d1b0-4aab-90c1-25323c0ca30e) |
| mint-compression-dist-xl.sh | :x: [more...](#2256f7ca-373c-4d5d-a581-51c62c5c601e) |
<a name="2256f7ca-373c-4d5d-a581-51c62c5c601e"></a>
### 8562-ee69310/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:31192
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ecb54fd32541:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 2 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 46 seconds
(8/14) Running minio-dotnet tests ... done in 37 seconds
(9/14) Running minio-go tests ... done in 49 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 6 seconds
(11/14) Running minio-js tests ... done in 1 minutes and 14 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 12 seconds
{
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty.",
  "status": "FAIL",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1172, in test_list_objects_with_1001_files\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 417, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1949, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2033, in main\n    test_list_objects_with_1001_files(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1174, in test_list_objects_with_1001_files\n    raise Exception(err)\nException: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n",
  "name": "minio-py:test_list_objects_with_1001_files",
  "args": {
    "bucket_name": "minio-py-test-ee00af6a-0d04-4e97-a67d-504f1e5f1a6e",
    "object_name": "d06c4db8-8d78-49bb-843c-f679bf851ef6_0 ~ d06c4db8-8d78-49bb-843c-f679bf851ef6_1000"
  },
  "function": "list_objects(bucket_name, prefix, recursive)",
  "duration": 104906
}

Executed 11 out of 14 tests successfully.
```
<a name="53c84b40-b36e-4f7e-92d5-8baea812ae20"></a>
### 8562-ee69310/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32210
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bbf35ad6a5a2:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-2anyya6f0c2zq9",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 364386,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15D96C1AD15C01D5, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="1f83be75-d1b0-4aab-90c1-25323c0ca30e"></a>
### 8562-ee69310/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:32046
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cc4a907a45b0:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 4 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 51 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 2 minutes and 12 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 16 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 9 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 2 seconds
{
  "args": {},
  "duration": 57704,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8562-ee69310

File: cmd/lock-rest-server.go [hunk 1]
```diff
@@ -127,6 +128,129 @@ func (l *lockRESTServer) RUnlockHandler(w http.ResponseWriter, r *http.Request)
...
+					}
+
+					// Call back to original server verify whether the lock is still active (based on name & uid)
+					expired, err := c.Expired(dsync.LockArgs{
+						UID:      nlrip.lri.UID,
+						Resource: nlrip.name,
+					})
+
+					if err != nil {
+						continue
```
Comment (klauspost) [inline cmd/lock-rest-server.go:214]: Maybe a `c.Close()` missing?

File: cmd/lock-rest-server.go [hunk 2]
```diff
@@ -127,6 +128,129 @@ func (l *lockRESTServer) RUnlockHandler(w http.ResponseWriter, r *http.Request)
...
+
+// Start lock maintenance from all lock servers.
+func startLockMaintenance() {
+	// Initialize a new ticker with a minute between each ticks.
+	ticker := time.NewTicker(lockMaintenanceInterval)
+	// Stop the timer upon service closure and cleanup the go-routine.
+	defer ticker.Stop()
+
+	// Start with random sleep time, so as to avoid "synchronous checks" between servers
+	time.Sleep(time.Duration(rand.Float64() * float64(lockMaintenanceInterval)))
```
Comment (klauspost) [inline cmd/lock-rest-server.go]: 👍 but you should grab the `ticker` *after* this, otherwise they will sync up again :)

File: cmd/local-locker.go [hunk 3]
```diff
@@ -202,6 +206,33 @@ func (l *localLocker) IsOnline() bool {
...
+		for _, entry := range lri {
+			if entry.UID == args.UID {
+				return false, nil
+			}
+		}
+	}
+	return true, nil
+}
+
+// Similar to removeEntry but only removes an entry only if the lock entry exists in map.
```
Comment (klauspost) [inline cmd/local-locker.go:225]: `// Caller must hold 'l.mutex' lock.`

File: cmd/lock-rest-server.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/lock-rest-server.go:214]: Correct

File: cmd/lock-rest-server.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/lock-rest-server.go]: Correct

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#fd2151fe-5671-4e4e-a73e-34bbf4f05c51) |
| mint-large-bucket.sh | :x: [more...](#3d5c1613-6e06-41ac-8754-b4ee7fc4eb3b) |
| mint-compression-dist-xl.sh | :x: [more...](#529bbf69-4caa-4f82-a030-af4a55588828) |
<a name="529bbf69-4caa-4f82-a030-af4a55588828"></a>
### 8562-59dbcf7/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:30625
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 05904d8df2f2:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 2 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 46 seconds
(8/14) Running minio-dotnet tests ... done in 45 seconds
(9/14) Running minio-go tests ... done in 50 seconds
(10/14) Running minio-java tests ... done in 57 seconds
(11/14) Running minio-js tests ... done in 52 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 7 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1172, in test_list_objects_with_1001_files\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 417, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1949, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2033, in main\n    test_list_objects_with_1001_files(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1174, in test_list_objects_with_1001_files\n    raise Exception(err)\nException: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n",
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty.",
  "args": {
    "bucket_name": "minio-py-test-05081e38-a25e-4bca-b821-0584d0828381",
    "object_name": "81c3981a-1636-4853-8cbd-2c5b3a20bc30_0 ~ 81c3981a-1636-4853-8cbd-2c5b3a20bc30_1000"
  },
  "function": "list_objects(bucket_name, prefix, recursive)",
  "name": "minio-py:test_list_objects_with_1001_files",
  "status": "FAIL",
  "duration": 100302
}

Executed 11 out of 14 tests successfully.
```
<a name="3d5c1613-6e06-41ac-8754-b4ee7fc4eb3b"></a>
### 8562-59dbcf7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31993
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 22375b5d1822:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 52 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 2 minutes and 22 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 15 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 11 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 2 seconds
{
  "args": {},
  "duration": 95436,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="fd2151fe-5671-4e4e-a73e-34bbf4f05c51"></a>
### 8562-59dbcf7/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:32199
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 86480d1180de:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-rv1ml9lv63vanq",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 364723,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15D9917829E0A159, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
Error: No such image: minio/minio:8562-59dbcf7

**Review by klauspost — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#1789650e-fcae-45ae-b459-d364d4a7fc78) |
| mint-xl.sh | :x: [more...](#d798c27f-2b0e-4b8d-b322-2833a704208d) |
| mint-large-bucket.sh | :x: [more...](#d65744b4-84ea-4c6d-b3bb-27b4b0b9882c) |
<a name="1789650e-fcae-45ae-b459-d364d4a7fc78"></a>
### 8562-bd7c3d5/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:30052
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3eb9447b32e6:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 4 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-t4zy6rtptmf00z",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 364262,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15DA7DFB053BA54E, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="d798c27f-2b0e-4b8d-b322-2833a704208d"></a>
### 8562-bd7c3d5/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:32304
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ce2943707cb6:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-2t4axwloov1gmu",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365245,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15DA7DFF5A6601B4, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="d65744b4-84ea-4c6d-b3bb-27b4b0b9882c"></a>
### 8562-bd7c3d5/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:32235
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f9468961a07a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 57 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 2 minutes and 19 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 11 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 12 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 4 seconds
{
  "args": {},
  "duration": 96077,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8562-bd7c3d5

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#f764f704-324e-4330-9665-0d1a005fb6a5) |
<a name="f764f704-324e-4330-9665-0d1a005fb6a5"></a>
### 8562-60b20b7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30095
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d9b813d03841:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 52 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 2 minutes and 29 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 15 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 22 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 49 seconds
{
  "args": {},
  "duration": 89407,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8562-60b20b7

**Review by poornas — APPROVED** [review]
LGTM

**Pull request merged**